docs/src/modules/ROOT/pages/using-timefold-solver/modeling-planning-problems.adoc

```diff
@@ -981,7 +980,214 @@ capacity: Annotated[int, PiggybackShadowVariable(shadow_variable_name="transport
 ----
 ====
 
-==== Shadow variable cloning
+[#declarativeShadowVariable]
+=== Declarative Shadow Variable
+
+Declarative shadow variables are the simplest way to calculate shadow variables.
+This approach allows access to all the fields and methods of the declaring class.
+
+[WARNING]
+====
+This feature is currently in xref:/upgrading-timefold-solver/backwards-compatibility.adoc#previewFeatures[preview] and may be subject to change.
+====
+
+To define a declarative shadow variable: Annotate the field with `@ShadowVariable` and specifying the name of a supplier method.
+Annotate the supplier method with `@ShadowSources`, listing all the planning variables (genuine or shadow) it depends on.
+
+Whenever any of the declared sources change, the supplier method is invoked, and its return value is set on all fields that reference it via @ShadowVariable.
+
+[tabs]
+====
+Java::
++
+[source,java,options="nowrap"]
+----
+@PlanningEntity
+public class Job {
+
+    @PlanningVariable
+    private LocalDate startDate;
+
+    private int durationInDays;
+
+    @ShadowVariable(supplierName="endDateSupplier")
+    private LocalDate endDate;
+
+    @ShadowSources("startDate")
+    public LocalDate endDateSupplier() {
+        if (startDate == null) {
+            return null;
+        } else {
+            return startDate.plusDays(durationInDays);
+        }
+    }
+}
+----
+
+Python::
++
+[source,python,options="nowrap"]
+----
+# Currently not supported in Python.
+----
+====
+
+In the example above, whenever the solver changes the `startDate` planning variable, the `endDateSupplier()` method is called.
+Its return value is assigned to the endDate field, as declared by the `@ShadowVariable(supplierName="endDateSupplier")` annotation.
+
+Some key considerations:
+
+- _Type consistency_: The return type of the supplier method must match the type of the field it is assigned to.
+- _Explicit dependencies_: All planning variables used in the supplier logic must be explicitly listed in `@ShadowSources`.
+Undeclared dependencies will not trigger updates, leading to stale or incorrect shadow values.
+
+
+==== @ShadowSources paths
+
+When using `@ShadowSources`, you must specify the paths to the variables that the supplier method depends on to compute its value.
+These paths must follow 1 of 3 syntactic forms:
+
+[cols="1,1,3", options="header"]
+|===
+|Form
+|Syntax Example
+|Description
+
+|Simple Variable Name
+|"variableName"
+|Refers to a variable (planning or shadow) on the same planning entity.
+
+|Chained Property Path
+|"a.b.c"
+|Refers to a supplier variable reachable via chained properties. Intermediate elements must be facts or variables; the final one must be a supplier variable.
+
+|Collection Element Access
+|"collectionVar[].varName"
+|Refers to a variable (`varName`) on each element in a collection (`collectionVar`) on the entity. The collection must not change during solving.
```

Comment:
Maybe you want to give an example of what this is for?
(FSR uses it to set the same `max(endTime)` to all those `varName` fields.)

Reply:
I tried to keep it more vague in the general explainer on purpose, as else you'd have to introduce domain specific elements (like you did). There is a full example a bit lower in the file.